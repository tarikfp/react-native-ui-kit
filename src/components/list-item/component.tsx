import PropTypes from "prop-types";
import React from "react";
import {
  ImageStyle,
  StyleProp,
  StyleSheet,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
  ViewStyle,
} from "react-native";
import { BaseColors, BaseFonts, BaseShadow } from "../../theme";
import ListItemButton from "./components/button";
import ListItemIcon from "./components/icon";
import ListItemImage from "./components/image";
import ListItemTitle from "./components/title";

const DEFAULT_LIST_ITEM_HEIGHT = 56;
export type Props = TouchableOpacityProps & {
  /**
   * List item height
   * @default 56
   */
  readonly height?: number;
  /**
   * List item style.
   */
  readonly style?: StyleProp<ViewStyle>;
  /**
   * Applies to shadow to the list item.
   * @default false
   */
  readonly hasShadow?: boolean;
  /**
   * Background color of the list item.
   * @default #FFFFFF
   */
  readonly backgroundColor?: string;
};

/**
 * ## Usage
 * ```js
 *     <ListItem height={72}>
 *       <ListItem.Icon name="light-bulb" type="Entypo" color="gold" />
 *       <ListItem.Title label="Find new ideas! " />
 *       <ListItem.Image
 *         source={{
 *           uri: "https://reactnative.dev/img/logo-og.png",
 *         }}
 *       />
 *       <ListItem.Icon
 *         alignment="end"
 *         name="arrow-right"
 *         type="MaterialCommunityIcons"
 *       />
 *     </ListItem>
 * ```
 */

const ListItem = ({
  height = DEFAULT_LIST_ITEM_HEIGHT,
  activeOpacity = 0.4,
  hasShadow = false,
  style,
  backgroundColor = "#FFFFFF",
  children,
  ...touchableOpacityProps
}: Props) => {
  const shadowStyle = hasShadow ? BaseShadow : undefined;

  return (
    <TouchableOpacity
      activeOpacity={activeOpacity}
      style={StyleSheet.flatten<ViewStyle>([
        styles.wrapper,
        { height, backgroundColor },
        shadowStyle,
        style,
      ])}
      {...touchableOpacityProps}>
      {React.Children.toArray(children)
        .filter((child) => child !== null && typeof child !== "boolean")
        .map((child, childInd) => {
          if (!React.isValidElement(child)) {
            return null;
          }

          let props:
            | {
                style?: StyleProp<ViewStyle> | StyleProp<TextStyle>;
                wrapperStyle?: StyleProp<ViewStyle>;
              }
            | undefined;

          if (child.type === ListItemTitle) {
            props = {
              wrapperStyle: {
                height,
                marginLeft: childInd !== 0 ? 8 : 0,
              },
              style: [BaseFonts.h3, child.props.style],
            };
          }

          if (child.type === ListItemIcon) {
            props = {
              style: [styles.icon, child.props.style],
            };
          }

          if (child.type === ListItemImage) {
            props = {
              style: [styles.image, child.props.style],
            };
          }

          if (child.type === ListItemButton) {
            props = {
              style: [styles.button, child.props.style],
            };
          }

          return React.cloneElement(child, props);
        })}
    </TouchableOpacity>
  );
};

ListItem.Icon = ListItemIcon;
ListItem.Image = ListItemImage;
ListItem.Title = ListItemTitle;
ListItem.Button = ListItemButton;

export default ListItem;

ListItem.propTypes = {
  height: PropTypes.number,
  hasShadow: PropTypes.bool,
  backgroundColor: PropTypes.string,
};

interface Styles {
  wrapper: ViewStyle;
  icon: TextStyle;
  image: ImageStyle;
  button: ImageStyle;
}

const styles = StyleSheet.create<Styles>({
  wrapper: {
    flexDirection: "row",
    borderRadius: 8,
    minWidth: "100%",
    alignItems: "center",
    paddingHorizontal: 12,
    paddingVertical: 8,
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: BaseColors.borderColor,
    backgroundColor: BaseColors.paper,
  },
  icon: {
    fontSize: 26,
    marginHorizontal: 4,
  },
  button: {
    marginHorizontal: 4,
  },
  image: {
    marginHorizontal: 4,
    width: 30,
    height: 36,
  },
});
