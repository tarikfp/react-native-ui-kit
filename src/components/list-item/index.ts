export { default as ListItem, Props as ListItemProps } from "./component";
