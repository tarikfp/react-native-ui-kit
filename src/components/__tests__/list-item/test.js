import * as React from "react";
import renderer from "react-test-renderer";
import { ListItem } from "../../list-item";

describe("tests list item component", () => {
  it("renders list item with title", () => {
    const tree = renderer
      .create(
        <ListItem>
          <ListItem.Title title="test" />
        </ListItem>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders list item with icon", () => {
    const tree = renderer
      .create(
        <ListItem>
          <ListItem.Icon name="arrow-left" type="MaterialCommunityIcons" />
        </ListItem>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders list item with image", () => {
    const tree = renderer
      .create(
        <ListItem>
          <ListItem.Image
            style={{ width: 40, height: 40 }}
            source={{ uri: "https://reactjs.org/logo-og.png" }}
          />
        </ListItem>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders list item with custom height", () => {
    const tree = renderer
      .create(
        <ListItem height={75}>
          <ListItem.Title title="Test" />
        </ListItem>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders list item with shadow", () => {
    const tree = renderer
      .create(
        <ListItem height={75} hasShadow>
          <ListItem.Title title="Test" />
        </ListItem>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders list item with custom backgroundColor", () => {
    const tree = renderer
      .create(
        <ListItem height={75} backgroundColor="orange">
          <ListItem.Title title="Test" />
        </ListItem>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders list item with custom style", () => {
    const tree = renderer
      .create(
        <ListItem style={{ paddingHorizontal: 20 }}>
          <ListItem.Title title="Test" />
        </ListItem>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
