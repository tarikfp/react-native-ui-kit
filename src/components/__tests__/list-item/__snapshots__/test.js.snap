// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests list item component renders list item with custom backgroundColor 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "orange",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 8,
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 75,
      "minWidth": "100%",
      "opacity": 1,
      "paddingHorizontal": 12,
      "paddingVertical": 8,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "justifyContent": "center",
        },
        Object {
          "height": 75,
          "marginLeft": 0,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(35, 47, 72, 1)",
            "fontSize": 16,
            "fontWeight": "500",
            "lineHeight": 24,
            "textAlign": "left",
          },
          Array [
            Object {
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
            },
            undefined,
          ],
        ]
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`tests list item component renders list item with custom height 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 8,
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 75,
      "minWidth": "100%",
      "opacity": 1,
      "paddingHorizontal": 12,
      "paddingVertical": 8,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "justifyContent": "center",
        },
        Object {
          "height": 75,
          "marginLeft": 0,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(35, 47, 72, 1)",
            "fontSize": 16,
            "fontWeight": "500",
            "lineHeight": 24,
            "textAlign": "left",
          },
          Array [
            Object {
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
            },
            undefined,
          ],
        ]
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`tests list item component renders list item with custom style 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 8,
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 56,
      "minWidth": "100%",
      "opacity": 1,
      "paddingHorizontal": 20,
      "paddingVertical": 8,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "justifyContent": "center",
        },
        Object {
          "height": 56,
          "marginLeft": 0,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(35, 47, 72, 1)",
            "fontSize": 16,
            "fontWeight": "500",
            "lineHeight": 24,
            "textAlign": "left",
          },
          Array [
            Object {
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
            },
            undefined,
          ],
        ]
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`tests list item component renders list item with icon 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 8,
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 56,
      "minWidth": "100%",
      "opacity": 1,
      "paddingHorizontal": 12,
      "paddingVertical": 8,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 0,
          "flexDirection": "row",
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#757575",
            "fontSize": 20,
          },
          Array [
            Object {
              "fontSize": 26,
              "marginHorizontal": 4,
            },
            undefined,
          ],
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      Û∞Åç
    </Text>
  </View>
</View>
`;

exports[`tests list item component renders list item with image 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 8,
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 56,
      "minWidth": "100%",
      "opacity": 1,
      "paddingHorizontal": 12,
      "paddingVertical": 8,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://reactjs.org/logo-og.png",
      }
    }
    style={
      Array [
        Object {
          "height": 36,
          "marginHorizontal": 4,
          "width": 30,
        },
        Object {
          "height": 40,
          "width": 40,
        },
      ]
    }
  />
</View>
`;

exports[`tests list item component renders list item with shadow 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 8,
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 75,
      "minWidth": "100%",
      "opacity": 1,
      "paddingHorizontal": 12,
      "paddingVertical": 8,
      "shadowColor": "rgba(45, 97, 185, 0.04)",
      "shadowOffset": Object {
        "height": 4,
        "width": 3,
      },
      "shadowOpacity": 1.25,
      "shadowRadius": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "justifyContent": "center",
        },
        Object {
          "height": 75,
          "marginLeft": 0,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(35, 47, 72, 1)",
            "fontSize": 16,
            "fontWeight": "500",
            "lineHeight": 24,
            "textAlign": "left",
          },
          Array [
            Object {
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
            },
            undefined,
          ],
        ]
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`tests list item component renders list item with title 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 8,
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 56,
      "minWidth": "100%",
      "opacity": 1,
      "paddingHorizontal": 12,
      "paddingVertical": 8,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "justifyContent": "center",
        },
        Object {
          "height": 56,
          "marginLeft": 0,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(35, 47, 72, 1)",
            "fontSize": 16,
            "fontWeight": "500",
            "lineHeight": 24,
            "textAlign": "left",
          },
          Array [
            Object {
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
            },
            undefined,
          ],
        ]
      }
    >
      test
    </Text>
  </View>
</View>
`;
